### ISSUE TITLE
Lora module layers - why float32?

### ISSUE THREAD

**Issue by cheald** [kohya-ss/sd-scripts#1156]
I was digging around in lora.py, and I noticed that the `lora_down` and `lora_up` layers created in LoRAModule don't specify a dtype, which results in their being allocated as float32. This is obviously good for precision, but I experimentally set it to bfloat16, and as expected, my RAM usage dropped (~11.5GB -> ~9.6GB) and my training speed went up substantially (~6.8 s/it -> 4.5 s/it - a 51% speed increase!), without an appreciable loss in training quality.

```diff
diff --git a/networks/lora.py b/networks/lora.py
index 948b30b..28a2bf8 100644
--- a/networks/lora.py
+++ b/networks/lora.py
@@ -62,14 +62,13 @@ class LoRAModule(torch.nn.Module):
             self.lora_down = torch.nn.Conv2d(in_dim, self.lora_dim, kernel_size, stride, padding, bias=False)
             self.lora_up = torch.nn.Conv2d(self.lora_dim, out_dim, (1, 1), (1, 1), bias=False)
         else:
-            self.lora_down = torch.nn.Linear(in_dim, self.lora_dim, bias=False)
-            self.lora_up = torch.nn.Linear(self.lora_dim, out_dim, bias=False)
+            self.lora_down = torch.nn.Linear(in_dim, self.lora_dim, bias=False, dtype=torch.bfloat16)
+            self.lora_up = torch.nn.Linear(self.lora_dim, out_dim, bias=False, dtype=torch.bfloat16)

         if type(alpha) == torch.Tensor:
             alpha = alpha.detach().float().numpy()  # without casting, bf16 causes error
```
This does obviously change the results during training somewhat, but it's not drastic, and I'm still getting good results.

Is there a reason that this wasn't already done? It seems like a pretty easy speed/RAM win for people who care about such things, and fp8/fp16/bf16 support is already present in other parts of the trainer.

**Comment by feffy380** [comment]
I think this is already be covered by `--full_fp16` and `--full_bf16`

**Comment by cheald** [comment]
This is embarrassingly correct - I have fp8_base turned on but somehow lost full_bf16 from my config file. Thanks for the sanity check.
